src/operator/fusion/fused_op.cu

```diff
@@ -0,0 +1,396 @@
+          std::string rhs, lhs;
+          rhs = variables[{node.inputs[0].node_id, node.inputs[0].index}];
+          if (act_type == "relu" ||
+              act_type == "sigmoid" ||
+              act_type == "tanh") {
+            lhs = variables[{node.inputs[1].node_id, node.inputs[1].index}];
+          } else {
+            lhs = variables[{node.inputs[2].node_id, node.inputs[2].index}];
+          }
+          code += "const auto " + var_name + " = backward_" + act_type +
+                  "(" + lhs + ", " + rhs + ");\n";
+
+          variables[{i, 0}] = var_name;
+          continue;
+        }
+        LOG(FATAL) << "Unrecognized op " + op_name;
+      }
+    } else {
+      LOG(FATAL) << "Encountered node with NULL source.";
+    }
+  }
+
+  int counter = 0;
+  for (const auto& entry : g.outputs()) {
+    const std::string& var = variables[{entry.node_id, entry.index}];
+    if (req[counter] == kWriteTo || req[counter] == kWriteInplace) {
+      code += "store(" + var + ", i, output" + std::to_string(counter) + ");\n";
+    } else if (req[counter] == kAddTo) {
+      code += "storeadd(" + var + ", i, output" + std::to_string(counter) + ");\n";
+    } else if (req[counter] == kNullOp) {
+      // NULL req, do not do anything
+    } else {
+      LOG(FATAL) << "Encountered unexpected req.";
+    }
+    ++counter;
+  }
+
+  this->code_ = code;
+}
+
+template <>
+void FusedOp::Forward<gpu>(const nnvm::NodeAttrs& attrs,
+                           const OpContext &ctx,
+                           const std::vector<TBlob> &inputs,
+                           const std::vector<OpReqType> &req,
+                           const std::vector<TBlob> &outputs) {
+  using namespace mshadow;
+  std::lock_guard<std::mutex> lock(my_mutex_);
+  CHECK_GE(outputs.size(), 1) << "There needs to be at least 1 output.";
+
+  std::vector<int> in_dtypes;
+  std::vector<int> out_dtypes;
+
+  size_t counter = 0;
+  for (const auto& blob : inputs) {
+    in_dtypes.push_back(blob.type_flag_);
+    initialized_ = initialized_ && (blob.type_flag_ == inputs_[counter].dtype);
+    inputs_[counter].dtype = blob.type_flag_;
+    ++counter;
+  }
+
+  counter = 0;
+  for (const auto& blob : outputs) {
+    out_dtypes.push_back(blob.type_flag_);
+    initialized_ = initialized_ && (blob.type_flag_ == outputs_[counter].dtype);
+    outputs_[counter].dtype = blob.type_flag_;
+    ++counter;
+  }
+
+  // Get compute capability of the current GPU
+  int dev_id = ctx.run_ctx.ctx.dev_id;
+  int cc_major = ComputeCapabilityMajor(dev_id);
+  int cc_minor = ComputeCapabilityMinor(dev_id);
+
+  initialized_ = initialized_ && cc_major == this->cc_major_;
+  initialized_ = initialized_ && cc_minor == this->cc_minor_;
+  this->cc_major_ = cc_major;
+  this->cc_minor_ = cc_minor;
+
+  initialized_ = initialized_ && (req == saved_reqs_);
+  saved_reqs_ = req;
+
+  if (!initialized_) {
+    this->GenerateCode(req);
```

Comment:
Could we wrap codegen into it's own function detail?

Reply:
ok